### PR TITLE
draft: feat: Unleash as env variable processor

### PR DESCRIPTION
**Pull request by ybenhssaien**
# Description

Make it possible to use unleash as an Environment variable in a config file (Yaml, XML, PHP)

> This is a simple not tested version stand alone (but the env var processor is tested in a real professional project)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [ ] Unit tests
- [ ] Spec Tests
- [ ] Integration tests / Manual Tests

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/DependencyInjection/UnleashEnvVarProcessor.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+use Unleash\Client\Unleash;
+
+final class UnleashEnvVarProcessor implements EnvVarProcessorInterface
+{
+    private Unleash $client;
+
+    public function __construct(Unleash $client)
+    {
+        $this->client = $client;
+    }
```
Comment (RikudouSage) [inline src/DependencyInjection/UnleashEnvVarProcessor.php:15]: Please use constructor promotion to be consistent with the rest of the code:

```suggestion
    public function __construct(private Unleash $client)
    {
    }
```

File: src/DependencyInjection/UnleashEnvVarProcessor.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    public function getEnv(string $prefix, string $name, \Closure $getEnv): bool
```
Comment (RikudouSage) [inline src/DependencyInjection/UnleashEnvVarProcessor.php:17]: Please import the `\Closure` to be consistent with the rest of the codebase.

File: src/Resources/config/services.yaml [hunk 3]
```diff
@@ -160,3 +160,8 @@ services:
       $cache: '@unleash.client.internal.cache'
     tags:
       - console.command
+
+  unleash.client.env_var_processor:
+    class: Unleash\Client\Bundle\DependencyInjection\UnleashEnvVarProcessor
+    arguments:
+      $client: '@unleash.client.unleash'
```
Comment (RikudouSage) [inline src/Resources/config/services.yaml:167]: Please add the tag here:

```suggestion
  unleash.client.env_var_processor:
    class: Unleash\Client\Bundle\DependencyInjection\UnleashEnvVarProcessor
    arguments:
      $client: '@unleash.client.unleash'
    tags:
      - container.env_var_processor
```

File: src/UnleashClientBundle.php [hunk 4]
```diff
@@ -19,6 +20,8 @@ public function build(ContainerBuilder $container): void
             ->addTag('unleash.client.strategy_handler');
         $container->registerForAutoconfiguration(BootstrapProvider::class)
             ->addTag('unleash.client.bootstrap_provider');
+        $container->registerForAutoconfiguration(UnleashEnvVarProcessor::class)
+            ->addTag('container.env_var_processor');
```
Comment (RikudouSage) [inline src/UnleashClientBundle.php:24]: Please remove this changes, the tag should be added in the yaml configuration.

**Review by RikudouSage — CHANGES_REQUESTED** [review]

File: src/DependencyInjection/UnleashEnvVarProcessor.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    public function getEnv(string $prefix, string $name, Closure $getEnv): bool
+    {
+        $value = $this->client->isEnabled($name);
+
+        // Env vars declared from yaml/xml files have string type
+        // 1 : Use unleash value first
+        // 2 : Retrieve the value of declared/default env var of unleash value is false or not retrieved
+        return $value || filter_var($getEnv($name), FILTER_VALIDATE_BOOL);
```
Comment (RikudouSage) [inline src/DependencyInjection/UnleashEnvVarProcessor.php:22]: Looking at this here, I don't think this is the correct approach... There should be some differentiation between Unleash returning false and the feature not existing. What happens here is that when the feature is evaluated to false a default is used, which IMO is not correct.

What do you think @sighphyre?

Comment (sighphyre) [reply src/DependencyInjection/UnleashEnvVarProcessor.php:22]: @RikudouSage Absolutely agree, PHP SDK does support default value as part of the isEnabled call, which is probably a better place for it. That should mean the fallback is only used when the toggle is missing, which is generally the correct flow for an SDK

Side note, if this isn't already using a context provider, I'd consider passing a context as well. Some toggle configurations, like a gradual rollout with a custom stickiness will always resolve (or should, I haven't checked) to false without a context, which is usually not what one wants

Comment (RikudouSage) [reply src/DependencyInjection/UnleashEnvVarProcessor.php:22]: @sighphyre This uses an instance provided by the Symfony dependency injection, meaning it already has the correct context.

@ybenhssaien Can you use the default value in the call to `isEnabled()`? It should be something like this:

```suggestion
        return $this->client->isEnabled($name, default: filter_var($getEnv($name), FILTER_VALIDATE_BOOL));
```

**Pull request closed without merging**